src/pal/src/loader/module.cpp

```diff
@@ -1339,29 +1342,59 @@ static HMODULE LOADLoadLibrary(LPCSTR shortAsciiName, BOOL fDynamic)
     // See if file can be dlopen()ed; this should work even if it's already loaded
     {
         // See GetProcAddress for an explanation why we leave the PAL.
-        PAL_LeaveHolder holder;
-        dl_handle = dlopen(shortAsciiName, RTLD_LAZY | RTLD_NOLOAD); 
-        if (!dl_handle)
+        PAL_LeaveHolder holder; 
+        
+        // Try to dlopen the module with appropriate variations on the name.
+        // P/Invokes are often declared with shortened versions of the real library
+        // name, such as without the extension or without the "lib" prefix.
+        // We try to open with such variations on the original.
+        int endsWithSuffix = FALSE, startsWithPrefixOrIsAbsolute = FALSE;
+        size_t shortAsciiNameLen = 0;
+        int i;
+        for (i = 1; i <= 4; i++) // four cases: original, with prefix, with suffix, with both prefix and suffix
         {
-            dl_handle = dlopen(shortAsciiName, RTLD_LAZY);
-        }
+            const char* filename = NULL;
+            switch (i)
+            {
+                case 1: // original name
+                    filename = shortAsciiName;
+                    break;
+
+                case 2: // with prefix prepended
```

Comment:
My thinking is that the two most likely cases people would use are *prefix*name*suffix* and name and that the plain name usage should be preferred.
So I think that the ideal order would be:
**prefix**name**suffix**
name
name**suffix**
**prefix**name (for the sake of completeness)

I would not try to detect whether the prefix or suffix are already there. It makes the code difficult to read and in case developers would follow the recommended way (plain name), the expected shared lib would be loaded at the first attempt. If they use something different, then they will pay the cost for that.

Then the implementation can be simplifed considerably, essentially each case being just a snprintf (except of the 2nd case) followed by attempt to load the library, without the need for the loop and switch.